src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -1,423 +1,370 @@
-
-    if (!std::stoi(localWeightStr)) { // maybe do this in some sort of initialization procedure instead of here? -Sam
-      /* Local weight hasn't been set */
-      int ret = cacheNode->set_attr(dpp, it->key, "localWeight", std::to_string(get_age())); 
-
-      if (ret < 0)
-	return {};
-    } else if (std::stoi(localWeightStr) < minWeight) {
-      minWeight = std::stoi(localWeightStr);
-      victimName = it->key;
-    }
-  }
+CacheBlock LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  if (entries_map.empty())
+    return {};
 
-  /* Get victim cache block */
-  CacheBlock victimBlock;
-  victimBlock.cacheObj.objName = victimName;
-  BlockDirectory blockDir;
-  blockDir.init(cct);
+  auto it = std::min_element(std::begin(entries_map), std::end(entries_map),
+			      [](const auto& l, const auto& r) { return l.second->localWeight < r.second->localWeight; });
 
-  int ret = blockDir.get_value(&victimBlock);
+  /* Get victim cache block */
+  CacheBlock victim;
+  victim.cacheObj.objName = it->second->key;
+  victim.cacheObj.bucketName = cacheNode->get_attr(dpp, victim.cacheObj.objName, "bucket_name", y); // generalize for other cache backends -Sam
+  victim.version = it->second->version; 
 
-  if (ret < 0)
+  if (dir->get(&victim, y) < 0) {
     return {};
-  #endif
-  CacheBlock victimBlock;
-  return victimBlock;
-}
-
-int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode) {
-  std::string key = "rgw-object:" + block->cacheObj.objName + ":directory";
-  optional_yield y = null_yield;
-  std::string localWeightStr = cacheNode->get_attr(dpp, block->cacheObj.objName, "localWeight", y); // change to block name eventually -Sam
-  int localWeight = -1;
+  }
 
-  if (!client.is_connected())
-    find_client(dpp, &client);
+  return victim;
+}
 
-  if (localWeightStr.empty()) { // figure out where to set local weight -Sam
-    optional_yield y = null_yield;
-    int ret = cacheNode->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(get_age()), y); 
-    localWeight = get_age();
+void LFUDAPolicy::shutdown() {
+  dir->shutdown();
+  
+  // call cancel() on the connection's executor
+  boost::asio::dispatch(conn->get_executor(), [c = conn] { c->cancel(); });
+}
 
-    if (ret < 0)
-      return -1;
-  } else {
-    localWeight = std::stoi(localWeightStr);
+int LFUDAPolicy::exist_key(std::string key, optional_yield y) {
+  if (entries_map.count(key) != 0) {
+    return true;
   }
 
-  int age = get_age();
+  return false;
+}
 
-  bool key_exists = true; //cacheNode->key_exists(dpp, block->cacheObj.objName) //TODO- correct this
-  if (key_exists) { /* Local copy */ 
-    localWeight += age;
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
```

Comment:
where is this method getting invoked, which will result in eviction of blocks if needed?